server/lib/amplitude.js

```diff
@@ -0,0 +1,245 @@
+const APP_VERSION = /^[0-9]+\.([0-9]+)\./.exec(require('../../package.json').version)[1];
+
+const GROUPS = {
+  email: 'fxa_email',
+  login: 'fxa_login',
+  registration: 'fxa_reg',
+  settings: 'fxa_pref',
+  sms: 'fxa_sms'
+};
+
+const ENGAGE_SUBMIT_EVENT_GROUPS = {
+  'force-auth': GROUPS.login,
+  signin: GROUPS.login,
+  signup: GROUPS.registration
+};
+
+const VIEW_EVENT_GROUPS = {
+  'force-auth': GROUPS.login,
+  settings: GROUPS.settings,
+  signin: GROUPS.login,
+  signup: GROUPS.registration,
+  sms: GROUPS.sms
+};
+
+const EMAIL_TYPES = {
+  'complete-reset-password': 'reset_password',
+  'complete-signin': 'login',
+  'verify-email': 'registration'
+};
+
+const NEWSLETTER_STATES = {
+  optIn: 'subscribed',
+  optOut: 'unsubscribed',
+};
+
+const EVENTS = {
+  'flow.reset-password.submit': {
+    group: GROUPS.login,
+    event: 'forgot_submit'
+  },
+  'settings.change-password.success': {
+    group: GROUPS.settings,
+    event: 'password'
+  },
+  'settings.clients.disconnect.submit': {
+    group: GROUPS.settings,
+    event: 'disconnect_device'
+  },
+  'settings.signout.success': {
+    group: GROUPS.settings,
+    event: 'logout'
+  },
+};
+
+const FUZZY_EVENTS = new Map([
+  [ /^flow\.([\w-]+)\.engage$/, {
+    isDynamicGroup: true,
+    group: eventCategory => ENGAGE_SUBMIT_EVENT_GROUPS[eventCategory],
+    event: 'engage'
+  } ],
+  [ /^flow\.[\w-]+\.forgot-password$/, {
+    group: GROUPS.login,
+    event: 'forgot_pwd'
+  } ],
+  [ /^flow\.[\w-]+\.have-account$/, {
+    group: GROUPS.registration,
+    event: 'have_account'
+  } ],
+  [ /^flow\.([\w-]+)\.submit$/, {
+    isDynamicGroup: true,
+    group: eventCategory => ENGAGE_SUBMIT_EVENT_GROUPS[eventCategory],
+    event: 'submit'
+  } ],
+  [ /^screen\.([\w-]+)$/, {
+    isDynamicGroup: true,
+    group: eventCategory => VIEW_EVENT_GROUPS[eventCategory],
+    event: 'view'
+  } ],
+  [ /^settings\.communication-preferences\.(optIn|optOut)\.success$/, {
+    group: GROUPS.settings,
+    event: 'newsletter'
+  } ],
+  [ /^([\w-]+).verification.success$/, {
+    isDynamicGroup: true,
+    group: eventCategory => eventCategory in EMAIL_TYPES ? GROUPS.email : null,
+    event: 'click'
+  } ]
+]);
+
+const NOP = () => {};
+
+const EVENT_PROPERTIES = {
+  [GROUPS.email]: mixProperties(mapEmailType, mapService),
+  [GROUPS.login]: mixProperties(mapEntrypoint, mapService),
+  [GROUPS.registration]: mixProperties(mapEntrypoint, mapService),
+  [GROUPS.settings]: mapEntrypoint,
+  [GROUPS.sms]: NOP
+};
+
+const USER_PROPERTIES = {
+  [GROUPS.email]: NOP,
+  [GROUPS.login]: mixProperties(mapUid, mapUtmProperties),
+  [GROUPS.registration]: mixProperties(mapUid, mapUtmProperties),
+  [GROUPS.settings]: mixProperties(mapUid, mapNewsletterState),
+  [GROUPS.sms]: NOP
+};
+
+module.exports = receiveEvent;
+
+function receiveEvent (event, data) {
+  if (! event || ! data) {
+    return;
+  }
+
+  const eventType = event.type;
+  let mapping = EVENTS[eventType];
+  let eventCategory;
```

Comment:
Is it alright if `eventCategory` is undefined?

Reply:
Yes.